finished issue 1237 - Save History

OliviaRuyinZhang opened the pull request:
Before submitting a final PR, please:

- [x] Run `npm test` on your machine
- [x] Run `npm run format` to keep code formatted consistently
- [x] Reference the issue you are fixing in at least one of your commit messages, as `Fixes #X`

1. Added a new file ./extension/history.js for indexedDB framework so that we can create table(s) to store history in indexedDB on browser. In the framework, I created the following methods: createTable, add, get, getAll, delete, and clearAll

2. Modified file extension/background/intentRunner.js: line 25 - 30 and line 363 to create a table to save all the history you say to firefox.

YufeiCui commented:
https://github.com/mozilla/firefox-voice/pull/1303

ianb commented on extension/history.js:
```diff
@@ -0,0 +1,129 @@
+    });
+  }
+
+  getAll(TBName) {
```
I believe this is the main method we'll want to be able to call, and we'll probably want to be able to get the items in sorted order (for us, sorted by timestamp, most recent first).

OliviaRuyinZhang commented:
Hi @ianb, new changes have been made upon request. 

In summary, here are the changes:

Under ``Extension/history.js``
1. Modified ``createTable()`` signatures to include version number
2. Modified ``reject()`` in all promises callback to raise an exception
3. Modified ``resolve()`` in all promises such that it either pass into a meaningful argunment or leavae it as blank
4. Modified ``getAll()`` method to sort by timestamp in the order of most recent first (descending) by default unless the user specify other sorting direction in the function parameter
5. Added some comment and docstring to clarify the usage of ``getAll()`` method
6. Modified variables name to make sure that they are all in ``camelCase``. 
7. Adjusted function signatures order so that they are consistent across different functions. i.e. now they are all in the order of ``tableName``, ``primaryKey``, ``object`` and etc.

Under ``Extension/Background/intentRunner.js``
1. Changed database's name from "myDB" to "voice" (line 25)
2. Changed ``createTable()`` parameters to adapte to the new signatures (line 29)
3. Added ``log.error()`` on promise reject under the function ``addIntentHistory()`` (line 384)


In the original issue #1237  you mentioned that you would like to have a UI to display all history to users in the future. If this is still part of the plan, could we have a separate issue to specify the requirenments for that and assign the new issue to @michael-mml ? We are a group of students from university of Toronto. We are doing a project course this semester surprvised by Anya Tafliovich. We would like to try to implement the UI part later.

ianb commented:
Great, looks good! I'll create a ticket about viewing history and tag @michael-mml in it (though I can't assign it to him unless he comments on that ticket).

ianb commented:
Followup in #1359
